plugin/code_actions.py

```diff
@@ -1,56 +1,124 @@
+    def send_request(self, on_response_recieved: 'Optional[Callable]' = None):
+        """ callback - hook with response as the first argument. """
+        session = session_for_view(self.view)
+        if not session or not session.has_capability('codeActionProvider'):
+            # the server doesn't support code actions, just return
+            return
+
+        params = self._get_code_action_params()
+        session.client.send_request(
+            Request.codeAction(params),
+            lambda response: self._handle_response(response, on_response_recieved))
+
+    def _handle_response(self, response, callback: 'Optional[Callable]' = None) -> None:
+        code_action = CodeAction(self.view)
+        if settings.show_bulb:
+            if len(response) > 0:
+                code_action.show_bulb()
+            else:
+                code_action.hide_bulb()
+
+        if callback is not None:
+            callback(response)
+
+    def show_bulb(self) -> None:
+        region = self.view.sel()[0]
+        flags = sublime.DRAW_NO_FILL | sublime.DRAW_NO_OUTLINE
+        self.view.add_regions('lsp_bulb', [region], 'markup.changed', 'Packages/LSP/icons/lightbulb.png', flags)
+
+    def hide_bulb(self) -> None:
+        self.view.erase_regions('lsp_bulb')
 
-    def run(self, edit, event=None):
-        client = client_for_view(self.view)
-        if client:
-            pos = get_position(self.view, event)
-            row, col = self.view.rowcol(pos)
-            point_diagnostics = get_point_diagnostics(self.view, pos)
-            params = {
-                "textDocument": {
-                    "uri": filename_to_uri(self.view.file_name())
-                },
-                "context": {
-                    "diagnostics": list(diagnostic.to_lsp() for diagnostic in point_diagnostics)
-                }
+    def _get_code_action_params(self):
+        region = self.view.sel()[0]
+        pos = region.begin()
+        point_diagnostics = get_point_diagnostics(self.view, pos)
+        return {
+            "textDocument": {
+                "uri": filename_to_uri(self.view.file_name())
+            },
+            "range": region_to_range(self.view, region).to_lsp(),
+            "context": {
+                "diagnostics": list(diagnostic.to_lsp() for diagnostic in point_diagnostics)
             }
-            params["range"] = region_to_range(self.view, self.view.sel()[0]).to_lsp()
-            if event:  # if right-clicked, set cursor to menu position
-                sel = self.view.sel()
-                sel.clear()
-                sel.add(sublime.Region(pos))
+        }
+
+
+class LspCodeActionListener(sublime_plugin.ViewEventListener):
+    @classmethod
+    def is_applicable(cls, _settings):
+        if settings.show_bulb:
+            return True
+        return False
+
+    def on_selection_modified_async(self):
+        self.fire_request()
+
+    def on_modified_async(self):
+        self.handle_modified_async()
```

Comment:
You explained this trigger with:
> It is needed to check if the code actions are valid after we made some changes.

What if we hide the bulb whenever the selection_modified_async fires?

Reply:
Yes, great suggestion.
I simplified that a bit, it this better 5079c9b?